File: src/main/java/dev/cbyrne/pufferfishmodloader/gradle/tasks/workspace/TaskDownloadAssets.java
package dev.cbyrne.pufferfishmodloader.gradle.tasks.workspace;

import dev.cbyrne.pufferfishmodloader.gradle.PufferfishGradle;
import dev.cbyrne.pufferfishmodloader.gradle.utils.HttpUtils;
import dev.cbyrne.pufferfishmodloader.gradle.utils.assets.AssetIndex;
import dev.cbyrne.pufferfishmodloader.gradle.utils.assets.AssetObject;
import org.gradle.api.DefaultTask;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.TaskAction;

import java.io.File;
import java.io.IOException;
import java.net.URL;

public class TaskDownloadAssets extends DefaultTask {
    @Input
    private AssetIndex assetIndex;
    @Internal
    private PufferfishGradle plugin;

    @TaskAction
    public void download() throws IOException {
        for (AssetObject object : assetIndex.getObjects().values()) {
            String path = object.getHash().substring(0, 2) + '/' + object.getHash();
            File file = new File(plugin.getCacheDir(), "assets/objects/" + path);
            URL url = new URL("http://resources.download.minecraft.net/" + path);
            HttpUtils.download(url, file, object.getHash(), 5, true);
        }
    }

    public PufferfishGradle getPlugin() {
        return plugin;
    }

    public void setPlugin(PufferfishGradle plugin) {
        this.plugin = plugin;
    }

    public AssetIndex getAssetIndex() {
        return assetIndex;
    }

    public void setAssetIndex(AssetIndex assetIndex) {
        this.assetIndex = assetIndex;

        for (AssetObject object : assetIndex.getObjects().values()) {
            String path = object.getHash().substring(0, 2) + '/' + object.getHash();
            File file = new File(plugin.getCacheDir(), "assets/objects/" + path);
            getOutputs().file(file);
        }
    }
}

File: src/main/kotlin/me/dreamhopping/pml/gradle/tasks/TaskDownloadAssets.kt
package me.dreamhopping.pml.gradle.tasks

import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.runBlocking
import me.dreamhopping.pml.gradle.mc.data.assets.AssetIndex
import me.dreamhopping.pml.gradle.utils.Hash
import me.dreamhopping.pml.gradle.utils.http.Downloader
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.TaskAction
import java.io.File

open class TaskDownloadAssets : DefaultTask() {
    @Input
    lateinit var index: AssetIndex
    @Input
    lateinit var output: String

    init {
        outputs.upToDateWhen { _ ->
            index.objects.any {
                val file = File(output, "objects/${it.value.hash.substring(0, 2)}/${it.value.hash}")
                file.exists() && Hash.sha1(file) == it.value.hash
            }
        }
    }

    @TaskAction
    fun download() {
        runBlocking {
            index.objects.map {
                val path = "${it.value.hash.substring(0, 2)}/${it.value.hash}"
                val file = File(output, "objects/$path")
                async { Downloader.download("http://resources.download.minecraft.net/$path", file, it.value.hash, silent = true) }
            }.awaitAll()
        }
    }
}